File: algorithm/sort/merge_sort/merge_sort.c
#include <stdio.h>
#include <stdlib.h>

void conquer(int arr[], size_t left, size_t mid, size_t right)
{
	size_t left_arr_idx = left;
	size_t right_arr_idx = mid + 1;
	int *sort_arr = (int *)malloc(sizeof(int) * (right + 1));

	int s_idx = left;
	while (left_arr_idx <= mid && right_arr_idx <= right)
	{
		if (arr[left_arr_idx] < arr[right_arr_idx]) // 오름차순 정렬
			sort_arr[s_idx] = arr[left_arr_idx++];
		else
			sort_arr[s_idx] = arr[right_arr_idx++];
		s_idx++;
	}

	if (left_arr_idx > mid)
	{
		for (int i = right_arr_idx; i <= right; i++, s_idx++)
			sort_arr[s_idx] = arr[i];
	}
	else
	{
		for (int i = left_arr_idx; i <= mid; i++, s_idx++)
			sort_arr[s_idx] = arr[i];
	}

	for (int i = left; i <= right; i++)
		arr[i] = sort_arr[i];

	free(sort_arr);
}

void merge_sort(int arr[], size_t left, size_t right)
{
	if (left >= right)
	{
		return;
	}

	// 분할
	size_t mid = (left + right) / 2;
	merge_sort(arr, left, mid);
	merge_sort(arr, mid + 1, right);

	// 정복
	conquer(arr, left, mid, right);
}

File: algorithm/sort/merge_sort/MergeSort.java
public class MergeSort {

	private static void conquer(Integer[] arr, Integer left, Integer mid, Integer right) {
		Integer leftArrIdx = left;
		Integer rightArrIdx = mid + 1;
		Integer[] sortArr = new Integer[arr.length];

		Integer sIdx = left;
		while (leftArrIdx <= mid && rightArrIdx <= right) {
			if (arr[leftArrIdx] < arr[rightArrIdx]) {
				sortArr[sIdx] = arr[leftArrIdx++];
			} else {
				sortArr[sIdx] = arr[rightArrIdx++];
			}
			sIdx++;
		}

		if (leftArrIdx > mid) {
			for (int i = rightArrIdx; i <= right; i++, sIdx++) {
				sortArr[sIdx] = arr[i];
			}
		} else {
			for (int i = leftArrIdx; i <= mid; i++, sIdx++) {
				sortArr[sIdx] = arr[i];
			}
		}

		for (int i = left; i <= right; i++)
			arr[i] = sortArr[i];
	}

	public static void mergeSort(Integer[] arr, Integer left, Integer right) {
		if (left >= right) {
			return;
		}

		//분할
		Integer mid = (left + right) / 2;
		mergeSort(arr, left, mid);
		mergeSort(arr, mid + 1, right);

		//병합
		conquer(arr, left, mid, right);
	}

	public static void main(String[] args) {
		Integer[] arr1 = {54, 1, 23, 9, 0, 7, 11};
		Integer[] arr2 = {1, 1, 1, 1, 1, 1, 1};
		Integer[] arr3 = {9, 8, 7, 6, 5, 4, 3, 2, 1, 0};
		mergeSort(arr3, 0, arr3.length - 1);

		for (int i = 0; i < arr3.length; i++) {
			System.out.println(arr3[i]);
		}
	}
}